query-grammar/src/lib.rs

```diff
@@ -4,12 +4,55 @@ mod user_input_ast;
 use combine::parser::Parser;
 
 pub use crate::occur::Occur;
-use crate::query_grammar::parse_to_ast;
+use crate::query_grammar::{parse_to_ast, SPECIAL_CHARS_NO_SPACE};
 pub use crate::user_input_ast::{UserInputAst, UserInputBound, UserInputLeaf, UserInputLiteral};
 
+#[derive(Debug)]
 pub struct Error;
 
 pub fn parse_query(query: &str) -> Result<UserInputAst, Error> {
     let (user_input_ast, _remaining) = parse_to_ast().parse(query).map_err(|_| Error)?;
     Ok(user_input_ast)
 }
+
+/// Removes special characters from a query. This also removes escaped characters.
+fn remove_special_chars_from_query(query: &str) -> String {
+    query.replace(SPECIAL_CHARS_NO_SPACE, "")
+}
+
+/// Attempts to do a strict parse of the query before removing all special
+/// characters and evaluating the query again.
+fn parse_query_leniently(query: &str) -> (UserInputAst, Option<Error>) {
+    match parse_query(query) {
+        Ok(uia) => (uia, None),
+        Err(e1) => match parse_query(&remove_special_chars_from_query(query)) {
+            Ok(uia) => (uia, Some(e1)),
+            // We should never reach this branch as the above query should always be valid
+            // with no special characters.
+            // TOOD: Is there a better way of handling this? Perhaps `query("")`?
+            Err(e2) => unreachable!(),
```

Comment:
How do will we prove or at least the right level of confidence that this is true (This is not true with the current PR)? Fuzzing?